Source/WebCore/rendering/style/PositionArea.cpp

```diff
@@ -0,0 +1,503 @@
+        return self == PositionAreaSelf::No ? CSSValueSpanBlockEnd : CSSValueSpanSelfBlockEnd;
+    case PositionAreaTrack::Center:
+        return CSSValueCenter;
+    case PositionAreaTrack::SpanAll:
+        return CSSValueSpanAll;
+    }
+
+    ASSERT_NOT_REACHED();
+    return CSSValueBlockStart;
+}
+
+static CSSValueID toInlineKeyword(PositionAreaTrack track, PositionAreaSelf self)
+{
+    switch (track) {
+    case PositionAreaTrack::Start:
+        return self == PositionAreaSelf::No ? CSSValueInlineStart : CSSValueSelfInlineStart;
+    case PositionAreaTrack::SpanStart:
+        return self == PositionAreaSelf::No ? CSSValueSpanInlineStart : CSSValueSpanSelfInlineStart;
+    case PositionAreaTrack::End:
+        return self == PositionAreaSelf::No ? CSSValueInlineEnd : CSSValueSelfInlineEnd;
+    case PositionAreaTrack::SpanEnd:
+        return self == PositionAreaSelf::No ? CSSValueSpanInlineEnd : CSSValueSpanSelfInlineEnd;
+    case PositionAreaTrack::Center:
+        return CSSValueCenter;
+    case PositionAreaTrack::SpanAll:
+        return CSSValueSpanAll;
+    }
+
+    ASSERT_NOT_REACHED();
+    return CSSValueInlineStart;
+}
+
+CSSValueID PositionAreaSpan::toKeyword() const
+{
+    switch (m_axis) {
+    case PositionAreaAxis::Horizontal:
+        ASSERT(m_self == PositionAreaSelf::No);
+        return toHorizontalKeyword(m_track);
+
+    case PositionAreaAxis::Vertical:
+        ASSERT(m_self == PositionAreaSelf::No);
+        return toVerticalKeyword(m_track);
+
+    case PositionAreaAxis::X:
+        return toXKeyword(m_track, m_self);
+
+    case PositionAreaAxis::Y:
+        return toYKeyword(m_track, m_self);
+
+    case PositionAreaAxis::Block:
+        return toBlockKeyword(m_track, m_self);
+
+    case PositionAreaAxis::Inline:
+        return toInlineKeyword(m_track, m_self);
+    }
+
+    ASSERT_NOT_REACHED();
+    return CSSValueLeft;
+}
+
+PositionArea::PositionArea(PositionAreaSpan span1, PositionAreaSpan span2)
+    : m_span1(span1)
+    , m_span2(span2)
+{
+}
+
+PositionArea PositionArea::fromKeyword(CSSValueID kw)
+{
+    auto maybeAxis = keywordToAxis(kw);
+    if (maybeAxis) {
+        // Keyword is axis unambiguous, second keyword is span-all
+        return PositionArea::fromKeywordPair(kw, CSSValueSpanAll);
+    }
+
+    // Keyword is axis ambiguous, it's repeated
+    return PositionArea::fromKeywordPair(kw, kw);
+}
+
+PositionArea PositionArea::fromKeywordPair(CSSValueID kw1, CSSValueID kw2)
+{
+    auto kw1Axis = keywordToAxis(kw1);
+    auto kw2Axis = keywordToAxis(kw2);
+
+    if (!kw1Axis && !kw2Axis) {
+        kw1Axis = PositionAreaAxis::Block;
+        kw2Axis = PositionAreaAxis::Inline;
+    } else if (!kw1Axis)
+        kw1Axis = getOppositeAxis(*kw2Axis);
+    else if (!kw2Axis)
+        kw2Axis = getOppositeAxis(*kw1Axis);
+
+    return {
+        { *kw1Axis, keywordToTrack(kw1), keywordToSelf(kw1) },
+        { *kw2Axis, keywordToTrack(kw2), keywordToSelf(kw2) }
+    };
+}
```

Comment:
If the parser is guaranteeing two values in the correct order, you can simplify your logic and all of your helper functions down to a single switch/case that constructs the PositionSpan object for each axis span from a keyword. Something like:

```
PositionSpan PositionSpan::fromKeyword(CSSValueID keyword)
{
switch (keyword) {
    case CSSValueLeft:
        return { Horizontal, Start, No };
    case CSSValueSpanLeft:
        return { Horizontal, SpanStart, No };
    case CSSValueRight:
        return { Horizontal, SpanEnd, No };
 etc.
}

PositionArea(CSSValueID blockOrXKeyword, CSSValueID inlineOrYKeyword)
: m_blockOrXSpan(PositionSpan::fromKeyword(blockOrXKeyword))
, m_inlineOrYSpan(PositionSpan::fromKeyword(inlineOrYKeyword))
{ }